File: PixelTrigger_BackgroundColor.py
'''
Get the closest RGB triplets for all triggers given a background rgb color

Inputs:
 
 backgroundcolor       3 element vector specifing the RGB values (between
                       0 and 255) of the background color (e.g. [200,200,200]
 show_visual           If set to true, a figure with colors that lead to 
                       different trigger values. The colors are chosen to 
                       minize the distance from the backgroundcolor. For each
                       plot the left color is the backgroundcolor and the
                       right is the trigger color. The bold trigger value
                       identifies the trigger value of your
                       backgroundcolor (best to choose a backgroundcolor
                       with trigger value 0)
 red_dim_samples       Number of samples along the red color dimension during
                       search. Values of 256 and higher lead to a full search
                       (really slow and often unnecessary). Default = 10

Outputs:
 
 rgb_trig_vals         dict with 256 entries: Each key belongs to a different 
                       trigger value while the entry for a key is the RGB triplet
                       with the closest possible color matching the background
                       


C.Postzich, 25.Dec.2021

'''


# Import Libraries
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.gridspec
import numpy as np


# Define Function
def PixelTrigger_BackgroundColor(backgroundcolor, show_visual, red_dim_samples = 10):

    print("Search for closest color to background for each trigger value:\n")

    all_comb = [bin(i) for i in range(256)]

    temp_blue = temp_green = []
    rgb_trig_vals = {}
    for i,x in enumerate(all_comb):
        temp_comb = list(x)
        temp_comb = temp_comb[2:]
        temp_comb = list('0'*max(0, 8-len(temp_comb))) + temp_comb

        temp_green = list('00000000')
        #temp_green[0::2] = temp_comb[0:4]
        temp_green[4:8] = temp_comb[4:8]
        temp_blue = list('00000000')
        #temp_blue[0::2] = temp_comb[4:8]
        temp_blue[4:8] = temp_comb[0:4]

        temp_closecol = np.zeros((256,4))
        for j,y in enumerate(all_comb):
            temp_comb2 = list(y)
            temp_comb2 = temp_comb2[2:]
            temp_comb2 = list('0'*max(0, 8-len(temp_comb2))) + temp_comb2
            #temp_green[1::2] = temp_comb2[0:4]
            temp_green[0:4] = temp_comb2[0:4]
            #temp_blue[1::2] = temp_comb2[4:8]
            temp_blue[0:4] = temp_comb2[4:8]
            temp_closecol[j,0:4] = get_closest_col([int(''.join(temp_green),2), int(''.join(temp_blue),2)], backgroundcolor, red_dim_samples)
        
        rgb_trig_vals.update({i: tuple([int(n) for _,n in enumerate(temp_closecol[temp_closecol[:,3].argmin(),0:3])])})
        disp_progress(i+1, len(all_comb))
    print("\n\n")

    #del rgb_trig_vals[[i for i,x in enumerate(rgb_trig_vals) if x[0][1] == backgroundcolor[1] and x[0][2] == backgroundcolor[2]][0]]

    if show_visual:

        outer_pad = 100
        left_pad = 8
        right_pad = 8
        upper_pad = 16
        lower_pad = 4
        background_trigval = rgb2triggervalue(backgroundcolor)

        def pick_patch(event):
            if event.guiEvent.num == 3:
                annot.set_visible(False)
                fig.canvas.draw_idle()
            else:
                if isinstance(event.artist, plt.Rectangle):
                    patch = event.artist
                    annot.set_position((event.mouseevent.x/1200, event.mouseevent.y/650))
                    annot.set_text(f'{[int(x*256) for x in patch.get_facecolor()[0:3]]}')
                    annot.set_visible(True)
                    fig.canvas.draw_idle()


        fig = plt.figure(figsize = (12.0, 6.5))
        ax = fig.add_axes([0,0,1,1], picker = True)
        ax.tick_params( bottom = False, left = False,labelbottom = False, labelleft = False)
        for i in range(256):
            x_len = 0.5*((1200/(18*1200)) - (left_pad+right_pad)/(1200-outer_pad))
            y_len = ((650/(15*650)) - (upper_pad+lower_pad)/650)
            x_offs = left_pad/(1200-outer_pad) + ((i % 18)*(1200-outer_pad))/(18*(1200))
            y_offs = 1 - (((int(np.floor(i/18))+1)*650)/(15*650) - lower_pad/650)
            ax.add_patch(plt.Rectangle((x_offs, y_offs), x_len, y_len, fill = True, edgecolor='black', facecolor = [y/256 for _,y in enumerate(backgroundcolor)],picker=True))
            ax.add_patch(plt.Rectangle((x_len+x_offs, y_offs), x_len, y_len, fill = True, edgecolor='black', facecolor = [y/256 for _,y in enumerate(rgb_trig_vals[i])],picker=True))
            if background_trigval == i:
                ax.text(x_len+x_offs,y_len+y_offs+2/650,i,fontdict = {'fontsize': 9,'fontweight': 'bold'}, color='r', ha='center')
            else:
                ax.text(x_len+x_offs,y_len+y_offs+2/650,i,fontdict = {'fontsize': 9}, ha='center')

        annot = ax.text(0, 0, "hello", bbox ={'facecolor':'yellow','alpha':0.6}, visible = False, picker = True)
        fig.canvas.mpl_connect("pick_event", pick_patch)
        plt.show()
        
    return rgb_trig_vals

def get_closest_col(bg, backgroundcolor, red_dim_samples):
    #eps = 0.0000000001
    #bkgd_norm = np.linalg.norm(backgroundcolor)
    start_val = int(np.mean(bg))
    col_dist_list = np.zeros((np.amin([255,start_val+red_dim_samples]) - np.amax([0, start_val-red_dim_samples]),4))
    for ind, c in enumerate(range(np.amax([0, start_val-red_dim_samples]), np.amin([255,start_val+red_dim_samples]), 1)):
        col_dist_list[ind,0:3] = [c] + bg
        #col_dist_list[ind,3] = 1 - np.dot(col_dist_list[ind,0:3],backgroundcolor) / np.max([eps, np.linalg.norm(col_dist_list[ind,0:3])]) / bkgd_norm
        col_dist_list[ind,3] = np.linalg.norm(col_dist_list[ind,0:3] - backgroundcolor)
    return(col_dist_list[col_dist_list[:,3].argmin(),])

def rgb2triggervalue(rgb):

    green_bin = list(bin(rgb[1]))
    green_bin = green_bin[2:]
    green_bin = list('0'*max(0, 8-len(green_bin))) + green_bin
    blue_bin = list(bin(rgb[2]))
    blue_bin = blue_bin[2:]
    blue_bin = list('0'*max(0, 8-len(blue_bin))) + blue_bin

    #trigger_bin = green_bin[0::2] + blue_bin[0::2]
    trigger_bin =  blue_bin[4:8] + green_bin[4:8]
    trigger_bin = "0b" + "".join(trigger_bin)
    return(int(trigger_bin,2))

def disp_progress(ind, full, barLength = 50):
    perc_done = float(ind) * 100 / full
    prog_done   = '-' * int(perc_done/100 * barLength)
    space  = ' ' * (barLength - len(prog_done))

    print('Progress: [%s%s] %d %% triggers evaluated' % (prog_done, space, perc_done), end='\r')


backgroundcolor = (128,128,128)
rgb_trig_vals = PixelTrigger_BackgroundColor(backgroundcolor,True)
print("\n".join("{}\t{}".format(k, v) for k, v in rgb_trig_vals.items()))

File: PixelTrigger_BackgroundColor.m
function rgb_trig_vals = PixelTrigger_BackgroundColor(backgroundcolor, varargin)

%% Get Possible Combinations of Green and Blue values corresponding to different triggers

%' Inputs:
% 
% backgroundcolor       3 element vector specifing the RGB values (between
%                       0 and 255) of the background color (e.g. [200,200,200]
% show_visual           If set to true, a figure with colors that lead to 
%                       different trigger values. The colors are chosen to 
%                       minize the distance from the backgroundcolor. For each
%                       plot the left color is the backgroundcolor and the
%                       right is the trigger color. The bold trigger value
%                       identifies the trigger value of your
%                       backgroundcolor (best to choose a backgroundcolor
%                       with trigger value 0)
%
%' Outputs:
% 
% rgb_trig_vals         255 x 4 list of RGB colors (column 1-3) and their 
%                       trigger values (column 4).
%
% C.Postzich, 25.Dec.2021


% Validate Function Inputs
validateattributes(backgroundcolor,{'numeric'},{'numel',3,'>=',0,'<',255,'nonnan'},'PixelTrigger_Colors','backgroundcolor')

if(isempty(varargin))
    show_visual = false;
else
    validateattributes(varargin{2},{'logical'},{'numel',1},'PixelTrigger_Colors','show_visual')
    show_visual = varargin{2};
end


all_combinations = dec2bin(0:255);

rgb_trig_vals = repmat([backgroundcolor(1),0,0],size(all_combinations,1),1);
for i = 1:size(all_combinations,1)
    temp_bin_g = '00000000';
    %temp_bin_g(1:2:8) = all_combinations(i,1:4);
    temp_bin_g(5:8) = all_combinations(i,1:4);
    temp_bin_b = '00000000';
    %temp_bin_b(1:2:8) = all_combinations(i,5:8);
    temp_bin_b(5:8) = all_combinations(i,5:8);
    temp_comb = zeros(size(all_combinations,1),3);
    for j = 1:size(all_combinations,1)
        %temp_bin_g(2:2:8) = all_combinations(j,1:4);
        temp_bin_g(1:4) = all_combinations(j,1:4);
        %temp_bin_b(2:2:8) = all_combinations(j,5:8);
        temp_bin_b(1:4) = all_combinations(j,5:8);
        temp_comb(j,1:2) = [bin2dec(temp_bin_g) bin2dec(temp_bin_b)];
        temp_diffcol = [transpose(0:255), repmat(temp_comb(j,1:2),256,1)];
        %[temp_comb(j,3), temp_comb(j,4)] = min(1 - dot(temp_diffcol, repmat(backgroundcolor,256,1),2) ./ vecnorm(backgroundcolor,1) ./ vecnorm(temp_diffcol,1,2));
        [temp_comb(j,3), temp_comb(j,4)] = min(sum((temp_diffcol - backgroundcolor).^2,2));
    end
    [~,idx] = min(temp_comb(:,3));
    
    rgb_trig_vals(i,1:3) = [temp_comb(idx,4)-1 temp_comb(idx,1:2)];
    rgb_trig_vals(i,4) = rgb2triggervalue(rgb_trig_vals(i,:));
end

% Plot Colors belonging to trigger values if visuals are wanted
if(show_visual)
    
    background_trigval = rgb2triggervalue(backgroundcolor);
    
    figure('pos', [112  57  1318  714])
    for i = 1:size(rgb_trig_vals,1)
        ax = axes('Position',[0.02+(0.98/18)*mod((i-1),18) 0.94-(0.98/15)*floor((i-1)/18) 0.04 0.04],'xtick',[],'ytick',[],'visible','off');
        patch([0 0.5 0.5 0],[0 0 1 1],backgroundcolor./256, 'tag',sprintf('Color (%d,%d,%d)',backgroundcolor(1), backgroundcolor(2), backgroundcolor(3)))
        patch([0.5 1 1 0.5],[0 0 1 1],rgb_trig_vals(i,1:3)./256)
        if(background_trigval == (i-1))
            text('parent',ax,  'String',num2str(rgb_trig_vals(i,4)),   'position',[0.5 1.25], 'FontSize',8, 'FontWeight','bold', 'Color','r',  'HorizontalAlignment','center')
        else
            text('parent',ax,  'String',num2str(rgb_trig_vals(i,4)),   'position',[0.5 1.25], 'FontSize',8,  'HorizontalAlignment','center')
        end
    end
    dcm = datacursormode;
    dcm.SnapToDataVertex = 'off';
    dcm.Enable = 'on';
    dcm.UpdateFcn = @displayColor;
    
end

end

% Callback function for verifying the blue and green values match the
% trigger value
function triggervalue = rgb2triggervalue(RGB)

green_bin = dec2bin(RGB(2));
green_bin = [repmat('0',1,8-length(green_bin)) green_bin];
blue_bin = dec2bin(RGB(3));
blue_bin = [repmat('0',1,8-length(blue_bin)) blue_bin];

%trigger_bin = [green_bin(1:2:8) blue_bin(1:2:8)];
trigger_bin = [green_bin(5:8) blue_bin(5:8)];
triggervalue = bin2dec(trigger_bin);

end

% Callback function for displaying color info in the figure
function txt = displayColor(~,info)
    txt = sprintf('RGB(%d,%d,%d)',info.Target.FaceColor(1)*256,info.Target.FaceColor(2)*256,info.Target.FaceColor(3)*256);
end
